Inferences/GeneralInduction.cpp

```diff
@@ -0,0 +1,517 @@
+  }
+  var++;
+
+  TermList t;
+  for (const auto& c : scheme.cases()) {
+    // rename variables in each case such that conclusion variables remain the same
+    auto rn = renameCase(c, scheme.inductionTerms(), var);
+    FormulaList* ll = FormulaList::empty();
+    for (auto& r : rn._recursiveCalls) {
+      auto f = createImplication(mainLit, sideLitQrPairs, r);
+      FormulaList::push(f, ll);
+      // save all free variables of the hypotheses -- these are used
+      // to mark the clauses as hypotheses and corresponding conclusion
+      if ((indhrw && mainLit->isEquality()) || (indmc && !mainLit->isEquality())) {
+        FormulaVarIterator fvit(f);
+        while (fvit.hasNext()) {
+          hypVars.insert(fvit.next());
+        }
+      }
+    }
+    auto right = createImplication(mainLit, sideLitQrPairs, rn._step);
+    Formula* left = 0;
+    if (FormulaList::isNonEmpty(ll)) {
+      left = JunctionFormula::generalJunction(Connective::AND, ll);
+    }
+    auto f = left ? new BinaryFormula(Connective::IMP, left, right) : right;
+    FormulaList::push(Formula::quantify(f), cases);
+  }
+
+  // create the substitution that will be used for binary resolution
+  // this is basically the reverse of the induction term map
+  ASS(FormulaList::isNonEmpty(cases));
+  RobSubstitution subst;
+  for (const auto& kv : scheme.inductionTerms()) {
+    ALWAYS(subst.match(TermList(kv.second, false), 0, TermList(kv.first), 1));
+  }
+  Formula* hypothesis = new BinaryFormula(Connective::IMP,
+    JunctionFormula::generalJunction(Connective::AND, cases),
+    Formula::quantify(createImplication(mainLit, sideLitQrPairs)));
+  // cout << *hypothesis << endl;
+
+  NewCNF cnf(0);
+  cnf.setForInduction();
+  Stack<Clause*> hyp_clauses;
+  Inference inf = NonspecificInference0(UnitInputType::AXIOM,_rule);
+  unsigned maxDepth = mainQuery.clause->inference().inductionDepth();
+  for (const auto& kv : sideLitQrPairs) {
+    maxDepth = max(maxDepth, kv.second.clause->inference().inductionDepth());
+  }
+  inf.setInductionDepth(maxDepth+1);
+  auto fu = new FormulaUnit(hypothesis,inf);
+  cnf.clausify(NNF::ennf(fu), hyp_clauses);
+  DHMap<unsigned,unsigned> rvs;
+  if ((indhrw && mainLit->isEquality()) || (indmc && !mainLit->isEquality())) {
+    // NewCNF creates a mapping from newly introduced Skolem symbols
+    // to the variables before Skolemization. We need the reverse of
+    // this, but we double check that it is a bijection.
+    // In other cases it may be non-bijective, but here it should be.
+    rvs.loadFromInverted(cnf.getSkFunToVarMap());
+  }
+
+  // Resolve all induction clauses with the main and side literals
+  auto resSubst = ResultSubstitution::fromSubstitution(&subst, 0, 1);
+  ClauseStack::Iterator cit(hyp_clauses);
+  while(cit.hasNext()){
+    Clause* c = cit.next();
+    for (const auto& v : hypVars) {
+      c->inference().addToInductionInfo(rvs.get(v));
+    }
+    auto qr = mainQuery;
+    qr.substitution = resSubst;
```

Comment:
Can we create 'qr' only once? It seems to be the same one for each iteration of this while loop.

Reply:
I made `mainQuery` and `sideLitQrPairs` non-references, so we copy the former and move the latter (since it is used only once) from outside and then we alter these before the binary resolution parts. I also perform the `Literal::complementaryLiteral` and `subst->apply` there since after that we don't use them anyways. Do you think this is reasonable?

Reply:
Looks good to me.